### PR TITLE
Add option to activate pulse counter peripheral manual on ESP32

### PR DESCRIPTION
**Pull request by m1ch**
## Description:
With this request it is possible to have more than 8 pulse_counter instances on an ESP32

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#988

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by m1ch** [comment]
@glmnet when can this PR be merged?

File: esphome/components/pulse_counter/sensor.py [hunk 1]
```diff
@@ -63,6 +63,27 @@ def validate_count_mode(value):
     return value
```
Comment (glmnet) [inline esphome/components/pulse_counter/sensor.py:65]: This doesn't work when compiling multi config files or ace/code validation as global doesn't get reset

**Review by glmnet — COMMENTED** [review]
Is this is kind of a big change? 

I'm wondering how difficult would be to instead deal with two components. I.e.  having a esp32 pulse counter and a interrupt pulse counter. 

Code will be simpler hence Code review would be easier, code selection would be handled by ESPHome (I.e. not #ifdef)

Not a request to do this way, just thinking out of the box 

A base class can be abstracted if necessary

**Comment by m1ch** [comment]
> Is this is kind of a big change?
> 
> I'm wondering how difficult would be to instead deal with two components. I.e. having a esp32 pulse counter and a interrupt pulse counter.
> 
> Code will be simpler hence Code review would be easier, code selection would be handled by ESPHome (I.e. not #ifdef)
> 
> Not a request to do this way, just thinking out of the box
> 
> A base class can be abstracted if necessary

@glmnet I like this idea! I can prepare the change of pulse_counter and the integration of interrupt_pulse_counter. This will be a breaking change, is this OK or do we need someone else's blessing before I start my work?

**Comment by ssieb** [comment]
I approve of splitting it into two components.  I would suggest that for a couple of releases, the "pulse_counter" platform still exists, but gives an error message directing to the new components and/or the documentation.

**Comment by m1ch** [comment]
@ssieb I suggest to keep the pulse_counter component for the software PC and add a component precice_pulse_counter with hardware-support on ESP32. I was thinking of issuing a warning for some time, if pulse_counter is used on ESP32. This way is the least impact, as I guess that the software PC is sufficient in a majority of cases.

**Pull request closed without merging**